### PR TITLE
Additional admin functions

### PR DESCRIPTION
**Pull request by jgknight**
### Summary
This PR finishes the previously stubbed out admin functionality for requesting and setting email address, confirmation status, and registration preference/status.

It also adds a check that was missing to the formatting screenname validation that the sn doesn't end in a space.

**Email Address**

- User can now set and update their email address using the AIM client
- Performs validation that it meets basic format user@host
- Performs validation that it is <= 320 characters
- Note that any waiting periods or confirmations etc are not required at this time

**Confirmation Status**

- User confirmation status is now reflected in the UserInfoFlags TLV. This can be seen when hovering mouse cursor over the user in your buddy list
- User can confirm their account if they have an email address set
- Note that any waiting periods or confirmations etc are not required at this time

**Registration Preference/Status**

- User can update their privacy preference in the AIM client and it is stored/reflected back to them

**Other changes**

- Added database migrations to add new columns for the email, confirmation, and reg pref data
- Added a clientInfoFlags member to the Session struct and added setters/getters so these flags can be updated more centrally. Note this also includes moving the away/unavailable flag setting logic to avoid a deadlock
- Updated unit tests

**Future changes**
Just wanted to mention that I'm intentionally not doing in this PR:

- Mgmt api - I need to figure out a good spot for the username and email address validation logic so that we can apply the same logic to new accounts via mgmt api, new accounts via disable_auth, and for updates from the AIM client

### Testing Done

Logging in as a brand new account, the account shows up as 'Unconfirmed Internet' in the buddy list
image

Attempting to format the screenname with one ending in a space results in an error
image

Attempting to confirm the account (My AIM > Edit Options > Confirm Account) without an email address set results in an error
image

Opening the update email address window, it starts out reporting the user has no email set

image

Attempting to enter an invalid email format results in an error
image

Setting the email to TestUser10@aol.com results in a success message. Note that as of this writing, the email change isn't actually pending and the user doesn't need to wait 72 hours. The email is immediately updated in the database, which can be seen if you open the update email address window again

image
image

Now with an email set, the account can be confirmed.  Attempting to reconfirm the account will tell the user it's already confirmed
image
image

Confirming the account sends a broadcast to all buddies with the updated userinfo tlv and now hovering over the name no longer shows the unconfirmed status
image

Resigning into AIM, the user continues to report as just 'Internet' not 'Unconfirmed Internet'

I also tested that setting Away/returning still works as I moved the logic for setting those flags.

Going to Preferences > Privacy, the toggle at the bottom can now be adjusted and saved
image

For the database migration I started with a database from retro-aim-server release version 0.8.0. Running the newer binary from my local build resulted in the database table having additional columns created and email address could be saved.

All unit tests pass.

### CODE REVIEW

File: foodgroup/admin.go [hunk 1]
```diff
@@ -68,16 +110,21 @@ func (s AdminService) InfoQuery(_ context.Context, sess *state.Session, frame wi
 		}
 	}
 
-	// wire.AdminTLVRegistrationStatus is used in the AIM Preferences > Privacy panel to control
-	// Allow users who know my e-mail address to find...
-	//	o Nothing about me - wire.AdminInfoRegStatusNoDisclosure
-	//	o Only that I have an account - wire.AdminInfoRegStatusLimitDisclosure
-	//	o My screen name - wire.AdminInfoRegStatusFullDisclosure
 	if _, hasRegStatus := body.TLVRestBlock.Slice(wire.AdminTLVRegistrationStatus); hasRegStatus {
```
Comment (mk6i) [inline foodgroup/admin.go:113]: Is it possible that all 3 of these items be requested in the same request?

File: foodgroup/admin.go [hunk 2]
```diff
@@ -158,6 +228,49 @@ func (s AdminService) InfoChangeRequest(ctx context.Context, sess *state.Session
...
+				},
+			}, nil
+		}
+		if err := s.accountManager.UpdateEmailAddress(e, sess.IdentScreenName()); err != nil {
+			return wire.SNACMessage{}, err
+		}
+		return replyMessage(wire.AdminTLVEmailAddress, e.Address), nil
+
+	} else if regStatus, hasRegStatus := body.TLVRestBlock.Slice(wire.AdminTLVRegistrationStatus); hasRegStatus {
+		regStatus := binary.BigEndian.Uint16(regStatus)
```
Comment (mk6i) [inline foodgroup/admin.go:252]: Let's use `body.TLVRestBlock.Uint16()` instead.

File: foodgroup/auth.go [hunk 3]
```diff
@@ -87,7 +90,18 @@ func (s AuthService) RegisterBOSSession(authCookie []byte) (*state.Session, erro
 		return nil, fmt.Errorf("user not found")
 	}
 
-	return s.sessionManager.AddSession(u.DisplayScreenName), nil
+	sess := s.sessionManager.AddSession(u.DisplayScreenName)
+	if sess == nil {
+		return nil, nil
+	}
```
Comment (mk6i) [inline foodgroup/auth.go:96]: I don't think this check really adds any value. If `sess` is nil, then a NPE will be raised from within `AddSession()`.

File: state/migrations/0005_user_settings.down.sql [hunk 4]
```diff
@@ -0,0 +1,17 @@
+CREATE TEMPORARY TABLE users_backup
```
Comment (mk6i) [inline state/migrations/0005_user_settings.down.sql:1]: Why a `TEMPORARY TABLE`?

File: state/session.go [hunk 5]
```diff
@@ -126,8 +151,14 @@ func (s *Session) UnsetIdle() {
 	s.idle = false
 }
 
-// SetAwayMessage sets the user's away message.
+// SetAwayMessage sets the user's away message, and
+// sets or clears the OServiceUserFlagUnavailable flag
 func (s *Session) SetAwayMessage(awayMessage string) {
+	if awayMessage != "" {
```
Comment (mk6i) [inline state/session.go:157]: The mutex lock needs to be at the top of the function body.

**Review by mk6i — CHANGES_REQUESTED** [review]
thanks for the PR, just a few small things

File: foodgroup/auth.go (hunk 3 above)
Comment (jgknight) [reply foodgroup/auth.go:96]: I'll take another look at it. iirc I added this to satisfy a unit test, `TestAuthService_RegisterBOSSession_SessionNotFound`

File: state/migrations/0005_user_settings.down.sql (hunk 4 above)
Comment (jgknight) [reply state/migrations/0005_user_settings.down.sql:1]: I'm not sure :) i stole it from `state/migrations/0002_hash_columns.down.sql`. I can change it to a regular table instead.

Comment (mk6i) [reply state/migrations/0005_user_settings.down.sql:1]: Ah bad copy pasta on my part :)

Comment (mk6i) [reply state/migrations/0005_user_settings.down.sql:1]: Fixed in f07c787

File: foodgroup/auth.go (hunk 3 above)
Comment (mk6i) [reply foodgroup/auth.go:96]: Removed the test in 5241751

File: foodgroup/admin.go (hunk 1 above)
Comment (jgknight) [reply foodgroup/admin.go:113]: For the InfoQuery it seems possible. I can update so it handles one or more in the same request.

For InfoChangeRequest it seems possible, although it's unclear how error handling would work in that case - if you had 2 items fail, how do you send the proper error back for both? And now that I'm rereading the docs, I might be using the wrong snac for errors - I'm using AdminErr but it seems to also work if I set it in an InfoChangeReply snac.

I'll do some more testing on this.

Comment (jgknight) [reply foodgroup/admin.go:113]: After testing and walking through libpurple

- We should only expect to receive a request for 1 item per snac. We can easily change this later if a client pops up that expects more than 1 item per snac for some reason.
- We should actually be sending errors back in AdminInfoReply or AdminInfoChangeReply snacs, not AdminErr. As I have it now, AdminErr works on AIM clients but pidgin specifically looks for those snac subgroup types to know if the change was a success or not.

So I'm going to leave what I have as-is but update the error responses to not use AdminErr.

**Comment by jgknight** [comment]
I've updated the error cases to send with the correct snac. I also noticed that pidgin/libpurple checks for the error TLV + a not-null URL TLV so that's why I'm sending it.

**Pull request merged**